Intersect.Server/Entities/Player.cs

```diff
@@ -1670,7 +1774,235 @@ public override void Warp(Guid newMapId, float newX, float newY, bool adminWarp
             }
         }
 
-        private void MapInstanceChanged(MapController newMap)
+        /// <summary>
+        /// Warps the player on login, taking care of instance management depending on the instance type the player
+        /// is attempting to login to.
+        /// </summary>
+        public void LoginWarp()
+        {
+            if (MapId == null || MapId == Guid.Empty)
+            {
+                WarpToSpawn();
+            }
+            else
+            {
+                if (!CanChangeToInstanceType(InstanceType, true, MapId))
+                {
+                    WarpToLastOverworldLocation(true);
+                }
+                else
+                {
+                    // Will warp to spawn if we fail to create an instance for the relevant map
+                    Warp(
+                        MapId, (byte)X, (byte)Y, (byte)Dir, false, (byte)Z, false, false, InstanceType, true
+                    );
+                }
+            }
+        }
+
+        /// <summary>
+        /// Warps the player to the last location they were at on the "Overworld" (empty Guid) map instance. Useful for kicking out of
+        /// instances in a variety of situations.
+        /// </summary>
+        /// <param name="fromLogin">Whether or not we're coming to this method via the player login/join game flow</param>
+        public void WarpToLastOverworldLocation(bool fromLogin)
+        {
+            Warp(
+                LastOverworldMapId, (byte)LastOverworldX, (byte)LastOverworldY, (byte)Dir, false, (byte)Z, false, false, MapInstanceType.Overworld, fromLogin
+            );
+        }
+
+        public void SendLivesRemainingMessage()
+        {
+            if (InstanceLives > 0)
+            {
+                PacketSender.SendChatMsg(this, Strings.Parties.InstanceLivesRemaining.ToString(InstanceLives), ChatMessageType.Party, CustomColors.Chat.PartyChat);
+            }
+            else
+            {
+                PacketSender.SendChatMsg(this, Strings.Parties.NoMoreLivesRemaining.ToString(InstanceLives), ChatMessageType.Party, CustomColors.Chat.PartyChat);
+            }
```

Comment:
The string should be ternaried here -- `(InstanceLives > 0 ? Strings.Parties.InstanceLivesRemaining : NoMoreLivesRemaining)`.ToString(InstanceLives).

Though why is `InstanceLives` even being passed to NoMorelivesRepaining?

Reply:
Just an oversight; fixed